php/class-wp-https-detection.php

```diff
@@ -0,0 +1,140 @@
+		add_filter( 'cron_request', array( $this, 'ensure_http_if_sslverify' ), PHP_INT_MAX );
+		add_action( 'parse_query', array( $this, 'verify_https_check' ) );
+	}
+
+	/**
+	 * Schedules a cron event to check for HTTPS support.
+	 */
+	public function schedule_cron() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::CRON_HOOK );
+		}
+	}
+
+	/**
+	 * Makes a request to find whether HTTPS is supported, and stores the result in an option.
+	 * If the request is a WP_Error, don't update the option.
+	 */
+	public function update_option_https_support() {
+		$https_support = $this->check_https_support();
+		if ( is_bool( $https_support ) ) {
+			update_option( self::HTTPS_SUPPORT_OPTION_NAME, $https_support );
+		}
+	}
+
+	/**
+	 * Makes a request to the home URL to determine whether HTTPS is supported.
+	 *
+	 * Returns true if the URL has the scheme of HTTPS, and it has the correct token.
+	 * The token is only to verify that the request is for the same site.
+	 * Also, in wp_remote_get(), 'sslverify' is true by default.
+	 *
+	 * @return boolean|WP_Error Whether HTTPS is supported, or a WP_Error.
+	 */
+	public function check_https_support() {
+		$this->query_var_number = wp_rand();
+		$response               = wp_remote_request(
+			add_query_arg(
+				self::REQUEST_QUERY_VAR,
+				$this->query_var_number,
+				home_url( '/', 'https' )
+			),
+			array(
+				'headers' => array(
+					'Cache-Control' => 'no-cache',
+				),
+			)
+		);
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		if ( false === strpos( wp_remote_retrieve_body( $response ), wp_hash( $this->query_var_number, 'nonce' ) ) ) {
+			return new WP_Error(
+				'invalid_https_validation_source',
+				__( 'There was an issue in the request for HTTPS verification. It might not have been from the same origin.', 'default' )
+			);
+		}
+
+		return 200 === wp_remote_retrieve_response_code( $response );
+	}
+
+	/**
+	 * If the 'cron_request' arguments include an HTTPS URL, this ensures sslverify is false.
+	 *
+	 * Prevents an issue if HTTPS breaks,
+	 * where there would be a failed attempt to verify HTTPS.
+	 *
+	 * @param array $request The cron request arguments.
+	 * @return array $request The filtered cron request arguments.
+	 */
+	public function ensure_http_if_sslverify( $request ) {
+		if ( 0 === strpos( $request['url'], 'https://' ) ) {
+			$request['args']['sslverify'] = false;
+		}
+		return $request;
+	}
+
+	/**
+	 * If the query has the query var, call die() with the hashed query var.
+	 *
+	 * This uses wp_die() instead of die(),
+	 * because it allows filtering the handler: 'wp_die_handler'.
+	 * This prevents stopping the unit tests.
+	 *
+	 * @param WP_Query $wp_query The query object.
+	 */
+	public function verify_https_check( $wp_query ) {
+		$query_var = $wp_query->get( self::REQUEST_QUERY_VAR );
+		if ( ! empty( $query_var ) ) {
+			wp_die( wp_hash( $query_var, 'nonce' ) ); // WPCS: XSS ok.
```

Comment:
If it's alright, this uses `wp_die()` instead of `die()`. That allows filtering the handler: `'wp_die_handler'`, which prevents stopping the unit tests.

Reply:
See https://github.com/xwp/pwa-wp/pull/16/files#r198711771